src/main/java/com/microsoft/sqlserver/jdbc/IdleConnectionResiliency.java

```diff
@@ -0,0 +1,455 @@
+/*
+ * Microsoft JDBC Driver for SQL Server Copyright(c) Microsoft Corporation All rights reserved. This program is made
+ * available under the terms of the MIT License. See the LICENSE file in the project root for more information.
+ */
+
+package com.microsoft.sqlserver.jdbc;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+class SessionRecoveryFeature {
+    private boolean connectionRecoveryNegotiated;
+    private int connectRetryCount;
+    private SQLServerConnection connection;
+    private SessionStateTable sessionStateTable;
+    private ReconnectThread reconnectThread;
+    private AtomicInteger unprocessedResponseCount = new AtomicInteger();
+    private boolean connectionRecoveryPossible;
+
+    /*
+     * Variables needed to perform a reconnect, these are not necessarily determined from just the connection string
+     */
+    private String loginInstanceValue;
+    private int loginNPort;
+    private FailoverInfo loginFailoverInfo;
+    private int loginLoginTimeoutSeconds;
+
+    SessionRecoveryFeature(SQLServerConnection connection) {
+        this.connection = connection;
+        reconnectThread = new ReconnectThread(connection);
+    }
+
+    boolean isConnectionRecoveryNegotiated() {
+        return connectionRecoveryNegotiated;
+    }
+
+    void setConnectionRecoveryNegotiated(boolean connectionRecoveryNegotiated) {
+        this.connectionRecoveryNegotiated = connectionRecoveryNegotiated;
+    }
+
+    int getConnectRetryCount() {
+        return connectRetryCount;
+    }
+
+    void setConnectRetryCount(int connectRetryCount) {
+        this.connectRetryCount = connectRetryCount;
+    }
+
+    SQLServerConnection getConnection() {
+        return connection;
+    }
+
+    void setConnection(SQLServerConnection connection) {
+        this.connection = connection;
+    }
+
+    ReconnectThread getReconnectThread() {
+        return reconnectThread;
+    }
+
+    SessionStateTable getSessionStateTable() {
+        return sessionStateTable;
+    }
+
+    void setSessionStateTable(SessionStateTable sessionStateTable) {
+        this.sessionStateTable = sessionStateTable;
+    }
+
+    boolean isConnectionRecoveryPossible() {
+        return connectionRecoveryPossible;
+    }
+
+    void setConnectionRecoveryPossible(boolean connectionRecoveryPossible) {
+        this.connectionRecoveryPossible = connectionRecoveryPossible;
+    }
+
+    int getUnprocessedResponseCount() {
+        return unprocessedResponseCount.get();
+    }
+
+    void resetUnprocessedResponseCount() {
+        this.unprocessedResponseCount.set(0);
+    }
+
+    void parseInitialSessionStateData(byte[] data, byte[][] sessionStateInitial) throws SQLServerException {
+        int bytesRead = 0;
+
+        // Contains StateId, StateLen, StateValue
+        while (bytesRead < data.length) {

```

Comment:
null check data - this causes ArrayIndexOutofBoundsException exception against Azure db.